Update HPC OnDemand instructions, beautify some things, add bitterroot, civet-like environment how-to (#27744)
idaholab/moose: modules/doc/content/application_development/build_system.md

## Overview

!style! halign=left
The MOOSE build system is a hierarchical Makefile based system. The idea is that no matter where
you are in the hierarchy of MOOSE applications/modules/framework, simply typing `make` will rebuild
everything that needs to be rebuilt. The basic way to build any MOOSE application is simply to do
a:

> !style-end!

Reviewer: FYI - if you wanted to reduce new line count, simply placing `!style halign=left` in front of the paragraph and not the full "start-to-end" syntax is sufficient to get the spacing right after a heading. Mentioning it in case you wanted to adjust as you're working through other parts of this review, but I will not require it, as this is good syntax as well.

Author: Oh! I didn't know that. Yes I'd much rather not need the style-end syntax.

  an advanced option and there are very few reasons to disable this. You will likely notice a
  dramatic slowdown if you set this to false.

## Build Methods

!style! halign=left
As mentioned above there are two environment variables that control what "type" of executable gets
created: `METHOD` and `METHODS`.  The executable that is ultimately created will have the `$METHOD`
as a suffix like: `something-$METHOD`. The valid options are:
!style-end!

- `opt`: (The default) Builds an optimized executable suitable for running calculations.
- `dbg`: An executable meant for use in a debugger (like `gdb` or `lldb`).  Will be very slow and
  not meant to be used on large problems
- `devel`: Something in-between `opt` and `dbg`.  Attempts to be faster but also does some more
  run-time checks (asserts and mooseAsserts)
- `oprof`: Not normally used.  Can be used for "code profiling": running your code with a utility
  like `oprof`, `gprof`, `vtune`, etc.

`METHODS` is used to specify the full set (space separated) of the above options that you will
 _ever_ want to used.  It is used in the build of `libMesh` to build a `libMesh` library for each of

>  the `METHODS`.

Reviewer: ```suggestion
 the `METHODS`, so that a rebuild of `libMesh` is not required every time a rebuild of MOOSE 
 with a new `METHOD` is performed.
```

## Build Optimization

> !style! halign=left

Reviewer: ```suggestion
!alert! note
```

Let's make this stand out a bit more since the section isn't strictly necessary.

This is advanced information meant to inform you of optimizations that happen within the MOOSE build
system. This is not necessary to understand how to use and build MOOSE.

> !style-end!

Reviewer: ```suggestion
!alert-end!
```

Tied to my previous comment.

There are two main build optimizations: "header symlinking" and "unity builds":

### Header Symlinking

!style! halign=left

> MOOSE and MOOSE-based application's include directories are organized into sub-directories for each

Reviewer: ```suggestion
MOOSE and MOOSE-based applications include directories are organized into sub-directories for each
```

system (such as `kernels`, `bcs`, `auxkernels`, etc.). There are currently ~30 different MOOSE
systems, meaning that each application could potentially have 30+ directories to search for include
files. In addition, applications can compose/use each other and use the physics modules. Each time

> you add another application/module you're gaining another 30+ include directories.

Reviewer: ```suggestion
you add another application/module, you're gaining another 30+ include directories.
```

180 paths!). This expansion of paths can lead to the compiler slowing down as each path needs to be
searched at compile time to find a header.

To combat this, during the build phase a directory named `build/header_symlinks` is created for each
application/module. All of the header files within that application/module are then symlinked into
that directory. This means that each application/module you add only adds _one_ include directory
(instead of 30), greatly speeding up the compilation process.

### Unity Builds

!style! halign=left
Within the subdirectories underneath the `src` directory of a typical MOOSE-based application you
will find definitions of multiple, closely-related objects in `.C` files.  For instance in
`src/kernels` you would be likely to find 10-15 Kernels that all inherit from `Kernel.h`.  However,
if we were to compile all of these files separately (as is commonly done) each of the `#include`
trees would have to be found and compiled for each one of these files individually. In addition,
each of these `.C` files would create a separate `.o` file that would ultimately need to be linked
into the final library / application.
!style-end!

> A different idea is to compile all of these files together, providing considerable savings for

Reviewer: ```suggestion
A different idea is to compile all of these files together as one file, providing considerable savings for
```

> evaluating the `#include`s and ultimately just producing _one_ `.o` file to be linked.  This can

Reviewer: ```suggestion
evaluating the `#include`s and ultimately just producing _one_ `.o` file to be linked.  This is known as a "unity build" and can
```

provide a huge savings in compile time. The following is showing timing for `make -j 16` for

> various sets of applications/modules/framework on the Falcon computer at Idaho National Laboratory.

Reviewer: ```suggestion
various sets of applications/modules/framework on the Falcon supercomputer at Idaho National Laboratory.
```

It is plain to see how much time is saved by using both header symlinking and unity builds.

!media media/application_development/falcon_compile_speed.png
        style=width:50%;float:right;margin-left:20px;
       caption=Compile speed on Falcon.

> The way we achieve this in MOOSE is using the idea of "Unity Builds". A Unity Build compiles
> multiple `.C` files together by create a `.C` file that `#include`s all of the other `.C` files. In

Reviewer: ```suggestion
The way we achieve a unity build in MOOSE is by compiling
multiple `.C` files together by create a `.C` file that `#include`s all of the other `.C` files. In
```

MOOSE this happens in a directory called `build/unity_src` in each application / module.

By default this behavior is +on+. To turn it off you can set `MOOSE_UNITY=false` in your
environment.  Further, an application can disable it permanently by putting `MOOSE_UNITY := false`
at the top of their Makefile.

> In addition you can also keep from unity building individual directories within your project. By

Reviewer: ```suggestion
In addition, you can also keep from using a unity build for individual directories within your project. By
```

> default `src` and `src/base` are not unity built because they generally contain a mix of objects

Reviewer: ```suggestion
default `src` and `src/base` are not built this way because they generally contain a mix of objects
```

Reducing awkward phrasing since what "this" is is clear from the previous sentence.

`METHODS` is used to specify the full set (space separated) of the above options that you will
 _ever_ want to used.  It is used in the build of `libMesh` to build a `libMesh` library for each of

>  the `METHODS`, so that a rebuild of `libMesh` is not required every time a rebuild of MOOSE.

Reviewer: ```suggestion
 the `METHODS`, so that a rebuild of `libMesh` is not required for every rebuild of MOOSE.
```

`src/kernels` you would be likely to find 10-15 Kernels that all inherit from `Kernel.h`.  However,
if we were to compile all of these files separately (as is commonly done) each of the `#include`
trees would have to be found and compiled for each one of these files individually. In addition,
each of these `.C` files would create a separate `.o` file that would ultimately need to be linked
into the final library / application.

A different idea is to compile all of these files together as one file, providing considerable
savings for evaluating the `#include`s and ultimately just producing _one_ `.o` file to be linked.
This is known as a "unity build" and can provide a huge savings in compile time. The following is
showing timing for `make -j 16` for various sets of applications/modules/framework on the [!ac](INL)
[!ac](HPC) Falcon supercompuer (now decommissioned). It is plain to see how much time is saved by
using both header symlinking and unity builds.

!media media/application_development/falcon_compile_speed.png
        style=width:50%;float:right;margin-left:20px;
       caption=Compile speed on Falcon.

> The way we achieve a unity build in MOOSE is by compiling multiple `.C` files together by create a

Reviewer: ```suggestion
The way we achieve a unity build in MOOSE is by compiling multiple `.C` files together by creating a
```

### Revision Generation

!style! halign=left
The build process can generate C++ header files containing repository versioning information for use
within your application. These header files are generated by default but can be disabled by setting

> a Make variable in your application's Makefile:

Reviewer: ```suggestion
a `Make` variable in your application's `Makefile`:
```

!style-end!

```
# must be set before the app.mk file is included
GEN_REVISION := no
include      $(FRAMEWORK_DIR)/app.mk
```

The Revision file contains both a version string (git hash) and a revision string containing more
detailed information or tagged information. Note that the name of your app will appear instead of
"MOOSE" for your individual application:

> 

Author: lol, this was never rendering properly.

ref: https://mooseframework.inl.gov/application_development/build_system.html#f257438f-88ba-4daf-887e-8e03cce37149

Reviewer: Those empty lines are super easy to forget....

#endif // MOOSE_REVISION_H
```

## MOOSE's build system

!style! halign=left

> MOOSE uses the tried and true Unix standard build tools,

Reviewer: ```suggestion
MOOSE uses the tried and true UNIX standard build tools,
```

[GNU Make](https://www.gnu.org/software/make/) and
[GNU Autoconf](https://www.gnu.org/software/autoconf/). These tools are ubiquitous, reliable, and

> stable, where many other build systems are not. We believe that usability of tool includes the build

Reviewer: ```suggestion
stable, where many other build systems are not. We believe that usability of a tool includes the build
```

optionally include "conf_var.mk" into our normal Makefile when it exists. This makes it so that
running "configure" is an optional step when building MOOSE or a MOOSE-based application.

### File list

- `framework/configure.ac` - This is the input file that is processed by `autoconf` to create the
  configure script. If you modify this file you will need to rerun `autoconf` to regenerate the
  `configure` script. If you defined a new variable (e.g. with AC_DEFINE), you will need to run
  `autoheader` then `autoconf`.

- `framework/include/base/MooseConfig.h.in` - This file is auto-generated from running `autoheader`.

- `framework/conf_var.mk.in` - This file is where you put your new expansion expressions. Typically
  these are in the form of "@VARIABLE@".

### Make Install

!style halign=left
MOOSE's build system has the ability to create installed binaries suitable for use with the conda
package manager, or installation on a shared computing resource such as Sawtooth. The `install`
target will copy the binary, and required libraries to a file tree based on the prefix you
configured with (set by the variable `PREFIX` or the `--prefix` argument).

> ```

Reviewer 2: ```suggestion
```bash
```

### Adding a new C-Preprocessor variable in MooseConfig.h

!style! halign=left
To add a new variable to MooseConfig.h, you'll first want to add the variable to configure.ac using
one of the variable definition macros such as "AC_DEFINE". After that, you'll run `autoheader`
followed by `autoconf`.
!style-end!

### Adding a new "make" variable

!style! halign=left
If you want to add an optional dependency to MOOSE that isn't already being handled by the libMesh
build system, you'll want likely want to use the "AC_SUBST" macros after adding the appropriate
logic to define those variables. You'll then edit the `conf_var.mk.in` file. Note: Since MOOSE
doesn't use automake, we don't generate a complete Makefile from running configure, rather we just
optionally include "conf_var.mk" into our normal Makefile when it exists. This makes it so that
running "configure" is an optional step when building MOOSE or a MOOSE-based application.
!style-end!

### File list

> - `framework/configure.ac` - This is the input file that is processed by "autoconf" to create the

Reviewer: ```suggestion
- `framework/configure.ac` - This is the input file that is processed by `autoconf` to create the
```

  configure script. If you modify this file you will need to rerun `autoconf` to regenerate the
  `configure` script. If you defined a new variable (e.g. with AC_DEFINE), you will need to run
  `autoheader` then `autoconf`.

- `framework/include/base/MooseConfig.h.in` - This file is auto-generated from running `autoheader`.

- `framework/conf_var.mk.in` - This file is where you put your new expansion expressions. Typically
  these are in the form of "@VARIABLE@".

### Make Install

!style! halign=left
MOOSE's build system has the ability to create installed binaries suitable for use with the conda
package manager, or installation on a shared computing resource such as Sawtooth. The `install`
target will copy the binary, and required libraries to a file tree based on the prefix you
configured with (set by the variable `PREFIX` or the `--prefix` argument).
!style-end!

```
cd moose
./configure --prefix=<installation path>
cd <application_dir>
make [make options]
make install
```

> Additionally each application may wish to make an one or more file structures available for end
> users to install into their own environment. This allows for end-users to install and run the

Reviewer: ```suggestion
Additionally each application may wish to make an one or more file structures available for end-users
to install into their own environment. This allows for end-users to install and run the
```

application test suite, examples, tutorials, or other inputs of interest. To customize what is
available for installation, the application developer will define the list of source directories
that can be installed by end users in their application Makefile. For example:

```
INSTALLABLE_DIRS := tests examples tutorials
```

!style! style=position:relative;top:-15px;left:5px;font-style:italic;font-size:small;
These paths should all be relative to the root of the application.
!style-end!

To give application developers the ability to customize the layout of the installed directories it's
possible to specify the destination path as well by using a "key/value" syntax. For example:

```
INSTALLABLE_DIRS := test/tests->tests examples/01->example_01 examples/02->example_02 tutorials
```

In this case, if the user chooses to install each of these directories into a writable workspace the
resulting directory structure will look like this:

```bash
$ ls -l
tests  example_01  example_02  tutorials
```

> !alert warning

Reviewer: ```suggestion
!alert! warning
```
Since we are using two lines after this alert, let's use the full syntax to ensure it is rendered properly

Make sure that installed paths do not include source code. This could result in an accidental export

> of information for those cleared for only binary access.

Reviewer: ```suggestion
of information for those cleared for only binary access.
!alert-end!
```

### Copying and Running Installed Inputs

!style! halign=left
Once an application has been installed, end users that have access to those binaries can install or

> copy those paths to their home directories and also run the tests using the test harness using

Reviewer: ```suggestion
copy those paths to their home directories and also run the tests with the MOOSE test harness using
```
